src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -115,48 +128,73 @@ library GPv2Encoding {
         bytes32 r = abi.decode(encodedOrder[140:], (bytes32));
         bytes32 s = abi.decode(encodedOrder[172:], (bytes32));
 
-        order.sellToken = tokens[order.sellTokenIndex];
-        order.buyToken = tokens[order.buyTokenIndex];
+        order.sellToken = tokens[sellTokenIndex];
+        order.sellTokenIndex = sellTokenIndex;
+        order.buyToken = tokens[buyTokenIndex];
+        order.buyTokenIndex = buyTokenIndex;
         order.kind = OrderKind((flags >> ORDER_KIND_BIT) & 0x1);
         order.partiallyFillable =
             (flags >> ORDER_PARTIALLY_FILLABLE_BIT) & 0x01 == 0x01;
 
-        bytes32 orderDigest;
-        bytes32 signingDigest;
-
         // NOTE: In order to avoid a memory allocation per call by using the
         // built-in `abi.encode`, we reuse the memory region reserved by the
-        // caller for the order result as input to compute the hash.
+        // caller for the order result as input to compute the hash, using the
+        // memory slot reserved for the order `owner` for the order type hash
+        // which is required by EIP-712 as a prefix to the order data.
+        // Furthermore structs are not packed in Solidity, and there is the
+        // order type hash prefix followed by the 9 order fields to hash for a
+        // total of `10 * sizeof(uint) = 320` bytes.
+        bytes32 orderDigest;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            // NOTE: Structs are not packed in Solidity, and there are 9 order
-            // fields to hash for a total of `9 * sizeof(uint) = 288` bytes.
-            orderDigest := keccak256(order, 288)
+            mstore(order, ORDER_TYPE_HASH)
```

Comment:
On top of being more explicit, it might actually be marginally cheaper to reserve a single field on the reusable struct once instead of doing a mem-copy for every order we process.

Reply:
The issue I see there is that the caller would have to set it once before calling decode. And then we would have the implicit requirement that the order memory needs to be prepped before calling `decode`. This should translate to a `PUSH; MSTORE` which is 5 gas/order.